src/main/java/hudson/remoting/Request.java

```diff
@@ -256,21 +256,22 @@ public RSP get() throws InterruptedException, ExecutionException {
             }
 
             public RSP get(long timeout, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException {
-                synchronized(Request.this) {
+                synchronized (Request.this) {
                     // wait until the response arrives
                     // Note that the wait method can wake up for no reasons at all (AKA spurious wakeup),
-                    long end = System.currentTimeMillis() + unit.toMillis(timeout);
-                    long now;
-                    while(response==null && (now=System.currentTimeMillis())<end) {
+                    long now = System.nanoTime();
+                    long end = now + unit.toNanos(timeout);
+                    while (response == null && (end - now > 0L)) {
```

Comment:
you need to account for overflow of the 64bit number.  It starts from an undetermined number - could easily be `Long.MAX_VALUE - 3` so end could be negative and now could be positive yet we still need to sleep.